### PR TITLE
Add a contributing guide

### PR DESCRIPTION
**Pull request by LukeWood**
We can eventually move this into governance if needed.

### CODE REVIEW

**Comment by LukeWood** [comment]
+haifeng and +matt as well - we will be reusing similar guides so it would be good to make sure I'm not missing anything here!

**Comment by bhack** [comment]
We are trying to standardize a template for this file in the ecosystem at:

https://github.com/tensorflow/community/pull/384

I would be nice If you could contribute the review and adopt it also in the Keras ecosystem

**Comment by saberkun** [comment]
Do you want to test all apis be compatible with all tf.distribute strategies? Do you want to build such testing infrastructure to help the community?

**Comment by LukeWood** [comment]
> Do you want to test all apis be compatible with all tf.distribute strategies?

ideally, but there may be some exceptions.  Right now the COCORecall metric isn’t compatible with the TPU strategy.

**Comment by LukeWood** [comment]
> Do you want to test all apis be compatible with all tf.distribute strategies? Do you want to build such testing infrastructure to help the community?

What would this look like?  Basically a parameterized test case where a param is test strategy?  I suppose it depends on if we can emulate distribution strategies without using underlying hardware

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+There may be
+several rounds of comments and code changes before the pull request gets
+approved by the reviewer.
+
+![Approval from reviewer](https://i.imgur.com/zgRziTt.png)
+
+### Step 6. Merging
+
+Once the pull request is approved, a `ready to pull` tag will be added to the
```
Comment (qlzh727) [inline CONTRIBUTING.md:48]: Since we are using github as source of truth, the 'ready to pull' is not used. The keras team member only need to merge the PR from github and doesn't need any copybara workflow to kickin.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+## Setup environment
+
+Setting up your KerasCV development environment is quite easy.  You simply
+need to run the following commands:
+
+```shell
+git clone https://github.com/YOUR_GITHUB_USERNAME/keras.git
+cd keras
+pip install ".[tests]"
```
Comment (qlzh727) [inline CONTRIBUTING.md:64]: what is ".[test]"? Usually we list all the required dependency in a requirements.txt file, and use pip install -r requirements.txt.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+Following these commands you should be able to run the tests using `pytest keras_cv`.
+Please report any issues running tests following these steps.
+
+## Run tests
+
+KerasCV is tested using [PyTest](https://docs.pytest.org/en/6.2.x/).
+
+### Run a test file
+
+To run a test file, run `pytest path/to/file`.
```
Comment (qlzh727) [inline CONTRIBUTING.md:77]: Assume this only works when the pwd is the root dir of keras_cv, otherwise the python import will error out.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+```
+pytest keras_cv/layers/preprocessing/cut_mix_test.py -k="label"
+```
+
+### Run all tests
+
+Running all of the tests in KerasCV is as easy as:
+```
+pytest keras_cv/
+```
```
Comment (qlzh727) [inline CONTRIBUTING.md:95]: do we also want to put some instruction about fix code format and linting?

**Review by qlzh727 — COMMENTED** [review]
Thanks for the PR.

**Comment by qlzh727** [comment]
> 

There are some existing testing utils in tensorflow/python/distribute for setting up various distribution strategies. Internally in google, the unit test can setup different hardware settings, eg multi-cpu/gpu/tpu and combinations, but it will be bit hard for any other local test env on user local. Only a few setting might work locally, eg mirror strategy with multi-cpu.

**Comment by qlzh727** [comment]
> We are trying to standardize a template for this file in the ecosystem at:
> 
> [tensorflow/community#384](https://github.com/tensorflow/community/pull/384)
> 
> It would be nice If you could contribute to the review and adopt it also in the Keras ecosystem

Thanks for the reference. For the code style and linting, is there a standard formatter used by tf community? We would love to use that to avoid reinventing the wheel.

File: CONTRIBUTING.md (hunk 1 above)
Comment (LukeWood) [reply CONTRIBUTING.md]: Good point - yeah I missed this.  Thanks

File: CONTRIBUTING.md (hunk 2 above)
Comment (LukeWood) [reply CONTRIBUTING.md:58]: .[tests] installs everything required to run unit tests.  You can see how this is handled in KerasTuner as well:
https://github.com/keras-team/keras-tuner/blob/master/setup.py#L40

So, installing:
```
pip install .
``` 
will install the list on line 30.  `pip install .[tests]` will also install pytest, flake8, isort, etc.

WDYT?

File: CONTRIBUTING.md (hunk 3 above)
Comment (LukeWood) [reply CONTRIBUTING.md]: Updated.

File: CONTRIBUTING.md (hunk 4 above)
Comment (LukeWood) [reply CONTRIBUTING.md:89]: Done

File: CONTRIBUTING.md (hunk 2 above)
Comment (LukeWood) [reply CONTRIBUTING.md:58]: + @haifeng-jin  + @mattdangerw to discuss best practices for managing dependencies as well.

I personally am a fan of the `setup.py` dependency management system.

Comment (mattdangerw) [reply CONTRIBUTING.md:58]: I am not sure! I don't really know enough about what is common practice in the python packaging world (or is this a legitimate source of contention?). It is nice to be able to split the testing and package requirements.

Some discussion:
https://stackoverflow.com/q/43658870
https://packaging.python.org/en/latest/discussions/install-requires-vs-requirements/

`install_requires` works and we have it now, so I am fine to leave as is unless we have a good reason to switch.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+### Step 2. Make code changes
+
+To make code changes, you need to fork the repository. You will need to setup a
+development environment and run the unit tests. This is covered in section
+"Setup environment".
+
+### Step 3. Create a pull request
+
+Once the change is ready, open a pull request from your branch in your fork to
+the master branch in [keras-team/keras](https://github.com/keras-team/keras).
```
Comment (mattdangerw) [inline CONTRIBUTING.md:23]: Switch link to keras-cv?

File: CONTRIBUTING.md (hunk 2 above)
Comment (LukeWood) [reply CONTRIBUTING.md:58]: Some extra context: the benefit of requirements is you can pin versions.

But at Google we are forced to update all dependencies anyways, so for Keras I think install_requires is superior.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+Once the pull request is approved, a team member will take care of merging.
+
+## Setup environment
+
+Setting up your KerasCV development environment is quite easy.  You simply
+need to run the following commands:
+
+```shell
+git clone https://github.com/YOUR_GITHUB_USERNAME/keras-cv.git
```
Comment (mattdangerw) [inline CONTRIBUTING.md:56]: I would maybe just move "forking and cloning" into the description above rather than these lines. There's not a universal way to clone (ssh vs https, etc).

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+Before making any changes, we recommend opening an issue (if one doesn't already
+exist) and discussing your proposed changes. This way, we can give you feedback
+and validate the proposed changes.
+
+If the changes are minor (simple bug fix or documentation fix), then feel free
+to open a PR without discussion.
+
+### Step 2. Make code changes
+
+To make code changes, you need to fork the repository. You will need to setup a
```
Comment (mattdangerw) [inline CONTRIBUTING.md:16]: I think we should link to a style guide or mention formatting somewhere here (fine if that's a follow up).

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+can use the following command to run all the tests in `cut_mix_test.py`,
+whose names contain `label`,
+
+```
+pytest keras_cv/layers/preprocessing/cut_mix_test.py -k="label"
+```
+
+### Run all tests
+
+Running all of the tests in KerasCV is as easy as:
```
Comment (mattdangerw) [inline CONTRIBUTING.md:86]: You use easy a couple times in this doc, which could be frustrating to someone whos environment is not working as it should. I think good general practice to avoid...

https://www.tensorflow.org/community/contribute/docs_style#principles_of_good_style

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
+## How to contribute code
```
Comment (mattdangerw) [inline CONTRIBUTING.md:1]: Should we include some discussion of filing an issue somewhere here? Could be a good place to capture any expectations we have (like a colab reproduction once we are fully packaged on pypi).

File: CONTRIBUTING.md (hunk 6 above)
Comment (LukeWood) [reply CONTRIBUTING.md:56]: Yeah, there isn't a universal way.  That being said, I think having a set of copy pasteable commands still serves value.  

A user who is concerned about the nuance between ssh/https cloning will probably be able to read the command and use their desired approach.  Thoughts?

File: CONTRIBUTING.md (hunk 8 above)
Comment (LukeWood) [reply CONTRIBUTING.md:86]: Good point, thanks for pointing this out to me.  I didn't think of that.

File: CONTRIBUTING.md (hunk 9 above)
Comment (LukeWood) [reply CONTRIBUTING.md:1]: Done - thanks!

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+You can run the unit tests for KerasCV by running:
+```
+pytest keras_cv/
+```
+
+### Formatting the Code
+In order to format the code you can use the `shell/format.sh` script.
+If this does not resolve the issue, try updating `isort` and `black`
+via `pip install --upgrade black` and `pip install --upgrade isort`.
```
Comment (bhack) [inline CONTRIBUTING.md:100]: This tools could give a different result locally then from the CI Action if don't specify on what version we rely in CI or we enforce in the repository like I have proposed with the pre-commit hooks PR.

File: CONTRIBUTING.md (hunk 6 above)
Comment (mattdangerw) [reply CONTRIBUTING.md:62]: If you want copy pastable, one option is to go with the `gh` command line tool? `gh repo fork keras-team/keras-cv --clone --remote` is a one liner that would handle forking, cloning and adding a remote in one go, and will trigger an auth flow if not already set up I think? It's also copy pastable without edits.

https://cli.github.com/manual/gh_repo_fork

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+### Run all tests
+
+You can run the unit tests for KerasCV by running:
+```
+pytest keras_cv/
+```
+
+### Formatting the Code
+In order to format the code you can use the `shell/format.sh` script.
```
Comment (mattdangerw) [inline CONTRIBUTING.md:98]: space after heading, use `##` not `###` (this isn't part of the pytest section)

File: CONTRIBUTING.md (hunk 2 above)
Comment (mattdangerw) [reply CONTRIBUTING.md:64]: If requirements.txt makes it easier to pin a system installed tensorflow that might be a reason to use it. I am not sure, but for running cloud TPU jobs with the setup.py we have today, I've needed to hack it up to stop it from installing a new tf version that does not have proper TPU support. No need to resolve here, but that's something I will keep looking into.

File: CONTRIBUTING.md (hunk 6 above)
Comment (LukeWood) [reply CONTRIBUTING.md]: Wow - that's a really nice command.  Thanks

File: CONTRIBUTING.md (hunk 11 above)
Comment (LukeWood) [reply CONTRIBUTING.md:101]: Updated

File: CONTRIBUTING.md (hunk 6 above)
Comment (bhack) [reply CONTRIBUTING.md]: I hope that we could point on a command that is already installed also on our Docker devel image (and our `.devcontainer`).

Then for extra things we could have a link to:
 https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository

File: CONTRIBUTING.md (hunk 7 above)
Comment (LukeWood) [reply CONTRIBUTING.md:20]: Let's do this in a follow up as that's a larger task.

File: CONTRIBUTING.md (hunk 11 above)
Comment (mattdangerw) [reply CONTRIBUTING.md:101]: also, maybe an intro to the tools we use up front. "We use flake8, black and isort for linting and formatting." See tuner

https://github.com/keras-team/keras-tuner/blob/master/CONTRIBUTING.md#code-style

Comment (LukeWood) [reply CONTRIBUTING.md]: Updated to follow the example in KerasTuner - thanks!

File: CONTRIBUTING.md (hunk 2 above)
Comment (haifeng-jin) [reply CONTRIBUTING.md:64]: I would suggest using `setup.py`. One caveat is that the current setting for version number [[Link](https://github.com/keras-team/autokeras/blob/master/setup.cfg#L1-L2)] may throw an error when installing with `pip install .` because it imports the package to get the version number before installing anything, which may encounter some not installed packages. For example, the user would have to install tensorflow before running `pip install .`. It is not a big deal though. The user can tell which package to install from the error.

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+```
+
+### Run all tests
+
+You can run the unit tests for KerasCV by running:
+```
+pytest keras_cv/
+```
+
+## Formatting the Code
```
Comment (haifeng-jin) [inline CONTRIBUTING.md:99]: Should we add something about the docstring styles?

File: CONTRIBUTING.md (hunk 6 above)
Comment (bhack) [reply CONTRIBUTING.md]: I think that we need at least to enforce the minimal for the linting/formatting tools at:
https://github.com/keras-team/keras-cv/blob/master/setup.py#L29

File: CONTRIBUTING.md (hunk 12 above)
Comment (LukeWood) [reply CONTRIBUTING.md:99]: Let's actually save style information for a fully fleshed out style guide.  I'll open a bug and write it up when I get some spare time.

File: CONTRIBUTING.md (hunk 10 above)
Comment (bhack) [reply CONTRIBUTING.md]: I think that we need at least to enforce the minimal for the linting/formatting tools at:
https://github.com/keras-team/keras-cv/blob/master/setup.py#L29

**Review by mattdangerw — APPROVED** [review]
Thanks for the edits! lgtm

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+"Setup environment".
+
+### Step 3. Create a pull request
+
+Once the change is ready, open a pull request from your branch in your fork to
+the master branch in [keras-team/keras-cv](https://github.com/keras-team/keras-cv).
+
+### Step 4. Sign the Contributor License Agreement
+
+After creating the pull request, the `google-cla` bot will comment on your pull
```
Comment (fchollet) [inline CONTRIBUTING.md:31]: "the `google-cla` bot will comment" -> unfortunately this is not true at this time. Do you know more about why it has been disabled? In any case, here it would be helpful to add a link to the CLA website that users can follow if they don't get a prompt on the PR (which is the current situation). Here: https://cla.developers.google.com/clas

**Review by fchollet — APPROVED** [review]

File: CONTRIBUTING.md (hunk 13 above)
Comment (LukeWood) [reply CONTRIBUTING.md:31]: Interesting, I didn't realize that the bot was disabled until now.  That's no good.  I added a comment linking to the CLA website.

Also, note that this is currently the in the keras-team/keras CONTRIBUTING.md guide as well @fchollet.  Maybe we need to update this as well.
